### PR TITLE
docs (external-data): Update readme to include new /register-session-url and /broadcast-signal endpoints

### PR DESCRIPTION
**Pull request by ghoshkaj**
[AB#3774](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3774)

Update the external-data example app readme to include the new endpoints introduced in https://github.com/microsoft/FluidFramework/pull/14565, https://github.com/microsoft/FluidFramework/pull/14592. 

Also make edits to make it cleared for partner teams to understand the full flow.

### CODE REVIEW

File: examples/hosts/app-integration/external-data/src/mock-customer-service/service.ts [hunk 1]
```diff
@@ -88,29 +88,6 @@ export async function initializeCustomerService(props: ServiceProps): Promise<Se
...
-	} catch (error) {
-		console.error(
-			formatLogMessage(
-				`Registering for data update notifications webhook with the external data service failed due to an error.`,
-			),
-			error,
-		);
-		throw error;
-	}
-
```
Comment (ghoshkaj) [inline examples/hosts/app-integration/external-data/src/mock-customer-service/service.ts:113]: This registration has moved to the body of the /register-session-url endpoint and does not really do anything without a externalTaskListId passed in. This should have been removed in the prior PRs.

File: examples/hosts/app-integration/external-data/README.md [hunk 2]
```diff
@@ -38,40 +37,30 @@ Next we need a customer service that functions as the intermediary between the E
...
+Find the details of the API in the [Customer Service README](./src/mock-customer-service/README.md).
 
-<img width="1356" alt="Scenario: Collaboration session gets an update from outside of Fluid" src="https://user-images.githubusercontent.com/6777404/216415477-14d0b193-29c9-48e5-8b6b-0a549a5dde58.png">
+**Fluid Service**
 
-<img width="1404" alt="Scenario: Collaboration session gets an update from outside of Fluid" src="https://user-images.githubusercontent.com/6777404/216417448-2a43db3e-12a2-48a6-b6d0-a6c4a95e27d1.png">
+_Driver Layer_
 
-<img width="1374" alt="Scenario: EXternal data is updated from within Fluid,if only authenticated users are able to write" src="https://user-images.githubusercontent.com/6777404/216417779-12861504-7909-489c-b7a2-4d75814a396f.png">
+The driver layer section is internal facing and anyone implementing this pattern need not be concerned with this. The changes in the driver layer are a new (and still under construction) `broadcast-signal` endpoint and the logic to relay the signal to the correct Fluid container when called.
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/README.md:50]: This was a bit hard to parse. Take these suggestions with a grain of salt, but they might make it a bit easier to read.
```suggestion
The driver layer section is internal facing, and anyone implementing this pattern need not be concerned with it. This layer is new (and still under construction). It exposes a `broadcast-signal` endpoint and the logic to relay the signal to the correct Fluid container when called.
```

File: examples/hosts/app-integration/external-data/README.md [hunk 3]
```diff
@@ -80,7 +69,15 @@ Task - This is the unit that can be edited and attributed to an author. It is al
 
 TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.
 
+TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.
+
+ITaskData - This is similar to Task except that it is the External Data Services model of the task.
+
+ITaskListData - This is similar to TaskList except that it is the External Data Services model of the taskList.
+
+<!-- prettier-ignore-start -->
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/README.md:78]: Is there a specific reason we have to ignore prettier here?

**Review by Josmithr — COMMENTED** [review]
I'm a little behind on some of the pattern changes, but I think I mostly followed what's going on. I left a couple of suggestions to enhance readability, but they represent my own opinions, so take them with a grain of salt :)

File: examples/hosts/app-integration/external-data/README.md (hunk 3 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/README.md:78]: It's complaining about the non-capitalized f for fluid in the SharedMap link (https://fluidframework.com/docs/data-structures/map/)

File: examples/hosts/app-integration/external-data/README.md (hunk 2 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/README.md:50]: Hmm you know, thinking about it -- it's not true that the implementor doesn't have to be concerned with it since they do need to explicitly call it within their implementation for the customer service. Maybe I should just leave this whole caveatout, except the bit where it's still a bit under construction and might change.

File: examples/hosts/app-integration/external-data/README.md (hunk 3 above)
Comment (alexvy86) [reply examples/hosts/app-integration/external-data/README.md:108]: Wait, we do have a markdown linter!? Why does it force us to put four spaces between the dash and the text in a list item and why doesn't it force blank lines before and after headers? 😡

Sorry for the tangent.

File: examples/hosts/app-integration/external-data/README.md [hunk 4]
```diff
@@ -18,59 +17,79 @@ In this case, the Fluid collaboration session serves as a "drafting surface" in
...
 
 This repo contains a service that mocks the external "source of truth" data server. This mock service offers a REST API collection and webhook interfaces in `./src/mock-external-data-service`. The API requests served by this "external" service are the following:
 
-1. POST `/register-for-webhook`: Register's the sender's URL to receive notifications when the external task-list data changes. Currently, the Customer Service registers its `/external-data-webhook` endpoint here to be called when data changes.
-2. GET `/fetch-tasks`: Fetches the task list from the external data store. Called by the Fluid client.
-3. POST `/set-tasks`: Updates external data store with new tasks list (complete override). Called by the Fluid client.
-4. POST `/debug-reset-task-list`: Resets the external data to its original contents. Called by the Fluid client.
+1. POST `/register-for-webhook?externalTaskListId=XXX`. Required body parameters: `url` (string). Registers the sender's URL to receive notifications when the external task-list data changes. Currently, the Customer Service registers its `/external-data-webhook` endpoint here to be called when data changes.
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:27]: Super nitty nit: I know it was me who suggested putting the task list id in the querystring. I think I should have suggested adding it to the body. Oh well, no biggie 😄 .

File: examples/hosts/app-integration/external-data/README.md [hunk 5]
```diff
@@ -18,59 +17,79 @@ In this case, the Fluid collaboration session serves as a "drafting surface" in
...
+Many services that would hold the "source of truth" data offer explicit commit style interfaces (e.g. via REST call or similar) which are not well suited to rapid updates.
 However, they often expose third-party integration via REST APIS for querying and manipulating data, as well as webhooks for watching updates to the data.
 
 This repo contains a service that mocks the external "source of truth" data server. This mock service offers a REST API collection and webhook interfaces in `./src/mock-external-data-service`. The API requests served by this "external" service are the following:
 
-1. POST `/register-for-webhook`: Register's the sender's URL to receive notifications when the external task-list data changes. Currently, the Customer Service registers its `/external-data-webhook` endpoint here to be called when data changes.
-2. GET `/fetch-tasks`: Fetches the task list from the external data store. Called by the Fluid client.
-3. POST `/set-tasks`: Updates external data store with new tasks list (complete override). Called by the Fluid client.
-4. POST `/debug-reset-task-list`: Resets the external data to its original contents. Called by the Fluid client.
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:31]: Don't feel too strongly about this last one. Keeping it here and saying it's for debugging purposes seems fine too.

File: examples/hosts/app-integration/external-data/README.md [hunk 6]
```diff
@@ -18,59 +17,79 @@ In this case, the Fluid collaboration session serves as a "drafting surface" in
...
+1.POST `/register-session-url`. Required body parameters: `containerUrl` (string), `externalTaskListId` (string). Creates an entry in the Customer Service of the mapping between the container and the external resource id. It then calls the External Service's `/register-for-webhook` endpoint to call it's own `/external-data-webhook?externalTaskListId=XXX` endpoint (details below), in order to notify the Fluid Service containers subscribed to that externalTaskListId of the corresponding change.
 
-### Echo Webhook Pattern
+2. POST `/external-data-webhook`. Required querystring parameters: `externalTaskListId`(string). Registered to be called by the External Data Service webhook when there's been a change to data upstream. On being called, the Customer Service calls the `/broadcast-signal` endpoint in the Fluid Service. Note that this is a route established by the Customer Service to be used exclusively as a subscription endpoint for the external data service's webhook, and should be considered a private implementation detail to the service. Customers may choose to implement this differently.
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:42]: Nit, rewording
```suggestion
2. POST `/external-data-webhook`. Required querystring parameters: `externalTaskListId`(string). This gets called by the External Data Service when there's been a change to the data and causes the Customer Service to in turn call the `/broadcast-signal` endpoint in the Fluid Service. Note that this is a route established by the Customer Service to be used exclusively as a subscription endpoint for the external data service's webhook and should be considered a private implementation detail to the service. Customers may choose to implement this differently.
```

File: examples/hosts/app-integration/external-data/README.md [hunk 7]
```diff
@@ -18,59 +17,79 @@ In this case, the Fluid collaboration session serves as a "drafting surface" in
...
+The `broadcast-signal` endpoint is new and still under construction. We will update it here once it is available to use.
 
-In the architecture so far, since the Customer Service is registered to the webhooks and listening for incoming changes, the clients do not have a way to know that there has been a change upstream. So the last piece of the puzzle here is an endpoint in the Fluid Service:
+1. POST `/broadcast-signal`. Required body parameters: `containerUrl` (string), `externalTaskListId` (string), `taskData`(ITaskData). On receiving this, the driver will broadcast a signal of `SignalType.RuntimeSignal` to the clients to alert them of an upstream change. This endpoint is called by the Customer Service to let the Fluid service know that there has been a change in the data. The body must contain the `conainerUrl` that is composed ot the `socketStreamUrl`, the `containerId` (sometimes known as the documentId), and the tenantId. In this way, it "echoes" the webhook from the External Data Service to the Customer Service. It is called by the Customer Service when it needs to notify the Fluid Clients that there has been a change to the data.
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:50]: Nit, rewording
```suggestion
1. POST `/broadcast-signal`. Required body parameters: `containerUrl` (string), `externalTaskListId` (string), `taskData`(ITaskData). This endpoint is called by the Customer Service when it needs the Fluid Service to notify the Fluid clients that there has been a change to the upstream data. The body must contain the `containerUrl`, from which we can extract a few pieces relevant to the Fluid service: the `socketStreamUrl`, the `containerId` (sometimes known as the `documentId`), and the `tenantId`. The Fluid service will broadcast a signal of type `SignalType.RuntimeSignal` to the clients to alert them of the change. In this way, it "echoes" the webhook call from the External Data Service to the Customer Service.
```

File: examples/hosts/app-integration/external-data/README.md [hunk 8]
```diff
@@ -18,59 +17,79 @@ In this case, the Fluid collaboration session serves as a "drafting surface" in
...
 
-1. POST `broadcast-signal`: Broadcasts a Signal to the clients to alert them of an upstream change. Called by the Customer Service to let the Fluid service know that there has been a change in the data.
+This example uses the example tinylicious driver to stub out what changes will be necessary in the odsp-driver. The prototype of the full signal and driver flow can be seen in this [`dev/external-data-prototyping` branch to main comparison](https://github.com/microsoft/FluidFramework/compare/main...dev/external-data-prototyping).
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:52]:
```suggestion
This example uses the tinylicious driver to stub out what changes will be necessary in the odsp-driver. The prototype of the full signal and driver flow can be seen in this [`dev/external-data-prototyping` branch to main comparison](https://github.com/microsoft/FluidFramework/compare/main...dev/external-data-prototyping).
```

File: examples/hosts/app-integration/external-data/README.md [hunk 9]
```diff
@@ -80,7 +99,15 @@ Task - This is the unit that can be edited and attributed to an author. It is al
 
 TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.
 
+TaskList - This can be compared to a "board" that holds all of the tasks. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:94]: "All tasks" makes me think of "all the tasks in the JIRA instance".
```suggestion
TaskList - This can be compared to a "board" that holds a list of related tasks; for example, the results of a particular query. It is the larger visible entity to show up in a component within the app. The app stores two types of TaskLists - a "draft" version and a "saved version". More on these below.
```

File: examples/hosts/app-integration/external-data/README.md
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/README.md:1]: Nit: the second image for the "Data changes on External Service" has a few typos in step 2.

The diff in Github isn't clear about which images are the final ones in the first flow :(

File: examples/hosts/app-integration/external-data/README.md (hunk 3 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/README.md]: Ooo you're right -- okay I was able to remove this!

**Review by alexvy86 — APPROVED** [review]
LGTM. I considered for a second if we'd prefer links to images stored in the repo next to the doc itself, but then remembered that @tylerbutler has pointed out in other PRs that we'd rather avoid committing binary files. Github user content seems like an ok alternative for images in the docs for our internal examples.
